### PR TITLE
Close open file descriptors and handle NBD_CMD_DISC correctly

### LINKED ISSUE #42: NBD server uses more and more memory and eventually crashes as clients keep connecting and disconnecting
I repeated the following commands:
```python
    c = new_nbd_client("localhost")
    c.close()
```
first with 0, then 1s, and finally with a **10s** delay between each iteration. In each case, as clients kept connecting and disconnecting, the memory usage of the NBD server grew linearly. In each case, after about **500 iterations**, when the server was using about **230Mb** of memory, it **crashed**:
```
nbd-tool: internal error, uncaught exception:
          Unix.Unix_error(Unix.EINVAL, "select", "")
```
I observed the same behaviour both with a custom Python client, (which doesn't close the TCP session during `c.close()`, because the server has to do it according to the NBD protocol), and versions `3.8` and `3.13` of the standard `nbd-client`. In the case of `nbd-client`, I repeated these commands:
```
sudo nbd-client localhost /dev/nbd2 -N ""
nbd-client -d /dev/nbd2
```

The reason for this issue might be that the NBD server does not handle `NBD_CMD_DISC` properly, it returns `EINVAL`.

I observed this issue with NBD server version: `2.1.3`, but it seems that `3.0.0` is also affected, but it behaves a bit differently. When its memory usage reaches 90Mb, it decreases to about 14Mb again. However, it still crashes after about 500 connections, but with a different error:
```
nbd-tool: internal error, uncaught exception:
          Unix.Unix_error(Unix.EMFILE, "accept", "")
```

### PR DESCRIPTION
**Pull request by gaborigloi**
Fixes #42

### CODE REVIEW

**Comment by gaborigloi** [comment]
Not sure whether I should also close `sock` in `cli/main.ml`.

Also, I used `Lwt_unix.close`, instead of `Server.close` (which does exactly the same thing), because this way I can also wrap `Server.connect`, which might also raise an exception. Maybe a more proper solution would be to use `Server.close` after `Server.connect`, as its implementation might change in the future and maybe it will do other cleanups as well?

**Comment by mseri** [comment]
I think it's better to use Server.close in this case. The closer the implementations of connect and close are, the better

**Comment by gaborigloi** [comment]
@mseri Then I'd have to wrap `Server.connect` with `Lwt_unix.close`, because that could also fail, but we don't have a `Server.t` at that point that we could call `Server.close` on. So it would make the code more complicated and less clean.

**Comment by gaborigloi** [comment]
@mseri I'll try to see if I can make use of `Server.close` in a simple way.

**Comment by thomassa** [comment]
You have the `channel` at that point though, so you can call `channel.close ()` (which is what `Server.close` does anyway). On the other hand, the implementation of `channel.close` is just `Lwt_unix.close fd` at present, so it all works out the same.

But once we have TLS it'll be different. Bear in mind that `Server` might need to initiate a "hard disconnect", which means shutting down the TLS connection (if any) and then the underlying TCP connection... so the `close` function of a TLS-protected `channel` will need to close down the TLS before doing anything else. (Note that `close` is not used at all anywhere in the code as it stands before this pull request.)

So I think we should try to find a clear place to draw the line separating the responsibilities of `Server` and `Main`: at present it's not obvious which really ought to do what.

We could consider making the channel's `close` function idempotent, but it would be cleaner to nest `Lwt.finalize` calls correctly so that `close` is always called exactly once.

**Comment by thomassa** [comment]
I think there are two broad approaches. One is based on nested `finalize` calls, something like
```
Lwt.finalize
  (fun () ->
     let channel = Nbd_lwt_unix.of_fd fd in
     Lwt.finalize
       (fun () ->
         Server.connect channel ()
         >>= fun (name, t) ->
         Lwt.finalize
           (fun () ->
             Block.connect filename
             >>= fun b ->
             Lwt.finalize
               (fun () -> Server.serve t (module Block) b)
               (fun () -> Block.disconnect b))
           (fun () -> Server.close t))
       (fun () -> channel.close ())
  (fun () -> Lwt_unix.close fd) in

...where `channel.close` does only any _extra_ clean-up that needs doing before closing the fd, and `Server.close` does only any _extra_ clean-up that needs to be done before closing the channel.

The other approach would use nested `try ... with _ -> close_something` instead: the shape of the code would be similar to the above (but using try-with instead of finalize) but each `close` function would do everything necessary, so `Server.close` would do all necessary clean-up including calling `channel.close`, and `channel.close` would do everything necessary including closing the fd.

I think I prefer the nested `finalize` approach.

**Comment by gaborigloi** [comment]
@thomassa There is also `Lwt.try_bind`, which might make things simpler: https://ocsigen.org/lwt/3.0.0/api/Lwt

**Comment by gaborigloi** [comment]
I've updated the code the use `Server.close` - should I also protect this line:
```ocaml
let channel = Nbd_lwt_unix.of_fd fd in
```
or can I assume that it doesn't throw an exception (that's currently true)?

**Comment by gaborigloi** [comment]
Also, should I close `socket` as well? Sometimes I get this error when I interrupt the server with CTRL + C, not sure if it's related:
```
nbd-tool: internal error, uncaught exception:
          Unix.Unix_error(Unix.EADDRINUSE, "bind", "")
```
However, when the server is interrupted with an exception, I don't get this error.

**Comment by mseri** [comment]
We need to add a signal handler. I think we should. the following out of context snippet shows how we did it in one of our daemons:

```ocaml
let start server = ..

let handle_shutdown handler () =
	Sys.set_signal Sys.sigterm (Sys.Signal_handle handler);
	Sys.set_signal Sys.sigint  (Sys.Signal_handle handler);
	Sys.set_signal Sys.sigpipe Sys.Signal_ignore

let () = ...
	(* Define the new signal handler *)
	let stop_handler signal = ..; exit 0 in
	let server = .. in
	let _ = (handle_shutdown stop_handler (); start server) in
	..
```

**Comment by mseri** [comment]
I think the `let channel` line should be ok. If it fails it is because the `fd` is unreadable or unavailable, I guess we can just crash in that case

**Comment by thomassa** [comment]
@mseri That `fd` is from one incoming connection, so if it's unreadable we should not crash the whole server, which might be handling other connections from other clients with no problem.
(It's fine for that one thread to die, though.)

**Comment by gaborigloi** [comment]
I think `let channel = Nbd_lwt_unix.of_fd fd in` doesn't do anything, it just constructs a record, so it shouldn't throw an exception.

**Comment by mseri** [comment]
I think `Cstruct` does some reading to construct the record

**Comment by gaborigloi** [comment]
I didn't manage to handle the signals in a convenient way by raising an exception from the handlers, and this might be a bug, see https://discuss.ocaml.org/t/lwt-how-to-catch-exceptions-raised-by-signal-handlers/565 and https://github.com/ocsigen/lwt/issues/451 .

**Comment by aantron** [comment]
We're looking into it on the Lwt end, though we can't promise a short time scale for fixing (or documenting?) it. Thanks for raising it.

**Comment by gaborigloi** [comment]
We've decided that we'll get this PR in first, which closes the fds, and then submit another PR with the signal handlers that uses a workaround. Now I also protect `let channel = Nbd_lwt_unix.of_fd fd in` with `Lwt.try_bind` (and not `Lwt.finalize`, to avoid double-closing the fd, which produces an exception). So can we merge this @thomassa @mseri ? (And should I also close the other server `socket`?)

**Comment by thomassa** [comment]
Yes, signal-handling can go in a separate pullreq.
It would seem tidy to close `sock` as the last action in thread `t`, though at present `t` runs `loop` which recurses for ever, so this would need to be done in an exception handler, or maybe `Lwt.finalize` such that once we have done `Lwt_unix.bind sock sockaddr` we can guarantee closing `sock`.

(As an aside: `let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in` is the place where eventually we'll want to specify the address (or addresses) rather than using `Unix.inet_addr_any` ...but that will be work for another day.)

File: cli/main.ml [hunk 1]
```diff
@@ -160,22 +160,41 @@ module Impl = struct
...
+        (fun () ->
+           let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in
+           Lwt_unix.bind sock sockaddr;
+           Lwt_unix.listen sock 5;
+           let rec loop () =
+             Lwt_unix.accept sock
+             >>= fun (fd, _) ->
+             (* Background thread per connection *)
+             let _ =
+               Lwt.try_bind
```
Comment (mseri) [inline cli/main.ml]: Can we extract both try_binds in separate functions? The code is getting lesser and lesser readable

Comment (mseri) [inline cli/main.ml]: Something like (untested and uncompiled):

```ocaml
let block_server channel = Lwt.try_bind
  (Server.connect channel)
  (fun (name, t) ->
    Lwt.finalize
      (fun () ->
        Block.connect filename
        >>= fun b ->
        Lwt.finalize
          (fun () -> Server.serve t (module Block) b)
          (fun () -> Block.disconnect b))
      (fun () -> Server.close t)
    )
    (fun _ -> channel.close ())
in
let prepare_channel_and_serve fd block_server = 
  Lwt.try_bind
    (fun () -> Lwt.wrap (fun () -> Nbd_lwt_unix.of_fd fd))
    (fun channel -> block_server channel)
in
... >>= fun (fd, _) -> prepare_channel_and_serve fd block_server
```

**Comment by gaborigloi** [comment]
@thomassa it doesn't throw an exception when we close the socket after construction:
```ocaml
# let sock = Lwt_unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, 123456) in Lwt_unix.close sock >>= fun () -> Lwt_io.print "closed\n" >>= fun () -> Lwt_unix.close sock >>= fun () -> Lwt_io.print "closed again\n";;                                                                                                                                               
closed
Exception: Unix.Unix_error (Unix.EBADF, "check_descriptor", "").
```

**Comment by gaborigloi** [comment]
@mseri done :)

**Comment by gaborigloi** [comment]
Thanks for the review!

**Comment by gaborigloi** [comment]
I realized that one issue with this PR is that when the block device doesn't exist the server keeps running, but each client connection will fail, instead of stopping the whole server.

**Comment by mseri** [comment]
Can you add an abort if the block device does not exist?

**Comment by gaborigloi** [comment]
Hmm, maybe it's not that important to stop the server if this could be a temporary error. But if it cannot be a temporary error, then we should stop it.

**Comment by mseri** [comment]
Probably yes. But I don't know. @djs55 do you have strong opinions on this?

**Pull request merged**